lib/gst/player/gstplayer.c

```diff
@@ -1121,6 +1165,514 @@ element_cb (GstBus * bus, GstMessage * msg, gpointer user_data)
+{
+  gint flags;
+
+  g_object_get (self->priv->playbin, "flags", &flags, NULL);
+  flags |= pos;
+  g_object_set (self->priv->playbin, "flags", &flags, NULL);
+}
+
+static void
+player_clear_flag (GstPlayer *self, gint pos)
+{
+  gint flags;
+
+  g_object_get (self->priv->playbin, "flags", &flags, NULL);
+  flags &= ~pos;
+  g_object_set (self->priv->playbin, "flags", &flags, NULL);
+}
+
+typedef struct
+{
+  GstPlayer *player;
+  GstPlayerMediaInfo  *info;
+}MediaInfoUpdatedSignalData;
+
+static gboolean
+media_info_updated_dispatch (gpointer user_data)
+{
+  MediaInfoUpdatedSignalData  *data = user_data;
+
+  g_signal_emit (data->player, signals[SIGNAL_MEDIA_INFO_UPDATED], 0, data->info);
+
+  return FALSE;
+}
+
+static void
+free_media_info_updated_signal_data (MediaInfoUpdatedSignalData *data)
+{
+  GST_DEBUG_OBJECT (data->player, "media_info %p", data->info);
+  g_object_unref (data->info);
+  g_slice_free (MediaInfoUpdatedSignalData, data);
+}
+
+/*
+ * emit_media_info_updated_signal:
+ *
+ * create a new copy of self->priv->media_info object and emits the newly created
+ * copy to user application. The newly created media_info will be unref'ed
+ * as part of signal finalize method.
+ */
+static void
+emit_media_info_updated_signal (GstPlayer *self)
+{
+  if (self->priv->dispatch_to_main_context) {
+    MediaInfoUpdatedSignalData *data = g_slice_new
+                                        (MediaInfoUpdatedSignalData);
+    data->player = self;
+    g_mutex_lock (&self->priv->lock);
+    data->info = gst_player_media_info_copy (self->priv->media_info);
+    g_mutex_unlock (&self->priv->lock);
+
+    GST_DEBUG_OBJECT (self, "media_info %p", data->info);
+
+    g_main_context_invoke_full (self->priv->application_context,
+        G_PRIORITY_DEFAULT, media_info_updated_dispatch,
+        data, (GDestroyNotify) free_media_info_updated_signal_data);
+  } else {
+    g_signal_emit (self, signals[SIGNAL_MEDIA_INFO_UPDATED], 0);
+  }
+}
+
+static GstCaps*
+get_caps (GstPlayer *self, gint stream_index, GType type)
+{
+  GstPad  *pad = NULL;
+  GstCaps *caps = NULL;
+
+  if (type == GST_TYPE_PLAYER_VIDEO_INFO)
+    g_signal_emit_by_name (G_OBJECT (self->priv->playbin),
+                          "get-video-pad", stream_index, &pad);
+  else if (type == GST_TYPE_PLAYER_AUDIO_INFO)
+    g_signal_emit_by_name (G_OBJECT (self->priv->playbin),
+                          "get-audio-pad", stream_index, &pad);
+  else
+    g_signal_emit_by_name (G_OBJECT (self->priv->playbin),
+                          "get-text-pad", stream_index, &pad);
+
+  if (pad) {
+    caps = gst_pad_get_current_caps (pad);
+    gst_object_unref (pad);
+  }
+
+  return caps;
+}
+
+static void
+gst_player_subtitle_info_update (GstPlayer *self,
+  GstPlayerStreamInfo *stream_info)
+{
+  GstPlayerSubtitleInfo *info = (GstPlayerSubtitleInfo*) stream_info;
+
+  if (stream_info->tags) {
+
+    if (info->language) {
+      g_free (info->language);
+      info->language = NULL;
+    }
```

Comment:
Maybe we should also do that if there are no tags anymore? Also the resetting to defaults in general